File: high_level/cpp/e2_direct_state_switch.cpp
#include <atomic>
#include <chrono>
#include <csignal>
#include <iostream>
#include <memory>
#include <string>
#include <thread>
#include <grpcpp/grpcpp.h>
#include "proto/grpc_service.grpc.pb.h"

using grpc_comm::ExecuteSequenceRequest;
using grpc_comm::ExecuteSequenceResponse;
using grpc_comm::gRPCService;
using grpc_comm::MotionSequence;

std::atomic<bool> g_interrupt {false};
void SignalHandler(int)
{
    g_interrupt.store(true);
}

class DirectStateSwitchClient
{
public:
    explicit DirectStateSwitchClient(const std::string& server_address)
        : server_address_(server_address)
    {
        auto channel = grpc::CreateChannel(server_address_, grpc::InsecureChannelCredentials());
        stub_ = gRPCService::NewStub(channel);
    }

    bool Run()
    {
        std::cout << "✓ Connected to server: " << server_address_ << std::endl;
        std::cout << "example 2: Direct State Switching Demo" << std::endl;

        ExecuteSequenceRequest request;
        auto* sequence = request.mutable_sequence();
        sequence->set_sequence_id("demo_state_switch");
        sequence->set_sequence_name("Directly State Machine Switching Demo");
        sequence->set_loop(false);

        for (const auto& id : {"passive", "stand_down", "stand_up", "x_legs", "balance_stand"}) {
            sequence->add_motions()->set_motion_id(id);
        }
        request.set_immediate_start(true);

        std::cout << "\nSequence is running... Press Ctrl+C to stop." << std::endl;

        ExecuteSequenceResponse response;
        grpc::ClientContext context;
        grpc::Status status;
        bool finished = false;
        bool cancelled = false;

        std::thread rpc_thread([&] {
            status = stub_->ExecuteSequence(&context, request, &response);
            finished = true;
        });
        rpc_thread.detach();

        while (!finished) {
            if (g_interrupt.exchange(false)) {
                cancelled = true;
                context.TryCancel();
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }

        if (!status.ok()) {
            if (status.error_code() == grpc::StatusCode::CANCELLED && cancelled) {
                std::cout << "Sequence cancelled by user." << std::endl;
            } else {
                std::cout << "RPC failed: " << status.error_message() << std::endl;
            }
            return false;
        }

        if (response.success()) {
            std::cout << "\nState switching demo executed successfully" << std::endl;
            std::cout << "  Execution ID: " << response.execution_id() << std::endl;
            std::cout << "  Message: " << response.message() << std::endl;
            std::cout << "  Number of motions: " << sequence->motions_size() << std::endl;
        } else {
            std::cout << "\nExecution failed: " << response.message() << std::endl;
        }
        return response.success();
    }

private:
    std::unique_ptr<gRPCService::Stub> stub_;
    std::string server_address_;
};

int main(int argc, char** argv)
{
    const std::string server_address = (argc > 1) ? argv[1] : "192.168.5.2:50051";

    // Register Ctrl+C handler
    std::signal(SIGINT, SignalHandler);

    DirectStateSwitchClient client(server_address);
    return client.Run() ? 0 : 1;
}

File: high_level/python/e2_direct_state_switch.py
#!/usr/bin/env python3
import grpc
import sys

from proto import grpc_service_pb2
from proto import grpc_service_pb2_grpc


def run_e2(server_address: str = "localhost:50051"):
    """
    states are switched by hands. User has to know which states are available under the current condition. Here we demonstrate a simple state switching sequence which is STAND_DOWN -> STAND_UP -> BALANCE_STAND -> DANCE0.
    
    This motion's id are "passive", "stand_down", "stand_up", "balance_stand", "dance0", etc. These have to be specified.
    Args:
        server_address: server address
    """
    # Establish connection
    channel = grpc.insecure_channel(server_address)
    stub = grpc_service_pb2_grpc.gRPCServiceStub(channel)
    print(f"✓ Connected to server: {server_address}")

    print("example 2: Direct State Switching Demo")
    print("STAND_DOWN -> STAND_UP -> BALANCE_STAND -> DANCE0")

    # Create motion sequence
    sequence = grpc_service_pb2.MotionSequence(
        sequence_id="demo_state_switch", sequence_name="Directly State Machine Switching Demo", loop=False
    )

    # Add motions to sequence in order
    # motion_ids = ["passive", "stand_down", "stand_up", "balance_stand", "dance0"]
    motion_ids = ["passive", "stand_down", "stand_up", "x_legs", "balance_stand"]
    for motion_id in motion_ids:
        motion = sequence.motions.add()
        motion.motion_id = motion_id

    # request define
    request = grpc_service_pb2.ExecuteSequenceRequest(sequence=sequence, immediate_start=True)

    call = stub.ExecuteSequence.future(request)
    print("\nSequence is running... Press Ctrl+C to stop.")

    try:
        while True:
            try:
                response = call.result(timeout=0.5)
                break
            except grpc.FutureTimeoutError:
                continue
    except KeyboardInterrupt:
        print("\nKeyboardInterrupt detected, cancelling execution...")
        call.cancel()
        try:
            call.result()
        except grpc.RpcError as e:
            if e.code() == grpc.StatusCode.CANCELLED:
                print("Sequence cancelled by user.")
            else:
                print(f"Sequence aborted: {e.code().name} - {e.details()}")
        return
    except grpc.RpcError as e:
        print(f"\nExecution failed: {e.code().name} - {e.details()}")
        return

    if response.success:
        print(f"\nState switching demo executed successfully")
        print(f"  Execution ID: {response.execution_id}")
        print(f"  Message: {response.message}")
        print(f"  Number of motions: {len(sequence.motions)}")
        print(f"  BPM: {sequence.bpm}")
    else:
        print(f"\nExecution failed: {response.message}")

    channel.close()


if __name__ == "__main__":
    server = sys.argv[1] if len(sys.argv) > 1 else "192.168.5.2:50051"
    run_e2(server)
